prefetcher/EIP/EIP.cc

```diff
@@ -0,0 +1,871 @@
+  l1i_timing_cache_table[l1i_cpu_id][set][index_cache].bere_line_addr = l1i_timing_mshr_table[l1i_cpu_id][index_mshr].bere_line_addr;
+  l1i_timing_cache_table[l1i_cpu_id][set][index_cache].accessed = l1i_timing_mshr_table[l1i_cpu_id][index_mshr].accessed;
+  l1i_invalid_timing_mshr_entry(line_addr);
+}
+
+// returns if accessed
+bool l1i_invalid_timing_cache_entry(uint64_t line_addr, uint64_t& bere_line_addr)
+{
+  uint32_t set = line_addr % L1I_SET;
+  uint32_t way = l1i_find_timing_cache_entry(line_addr);
+  assert(way < L1I_WAY);
+  l1i_timing_cache_table[l1i_cpu_id][set][way].valid = false;
+  bere_line_addr = l1i_timing_cache_table[l1i_cpu_id][set][way].bere_line_addr;
+  return l1i_timing_cache_table[l1i_cpu_id][set][way].accessed;
+}
+
+void l1i_access_timing_entry(uint64_t line_addr)
+{
+  uint32_t index = l1i_find_timing_mshr_entry(line_addr);
+  if (index < L1I_TIMING_MSHR_SIZE) {
+    if (!l1i_timing_mshr_table[l1i_cpu_id][index].accessed) {
+      l1i_timing_mshr_table[l1i_cpu_id][index].accessed = true;
+    }
+    return;
+  }
+  uint32_t set = line_addr % L1I_SET;
+  uint32_t way = l1i_find_timing_cache_entry(line_addr);
+  if (way < L1I_WAY) {
+    l1i_timing_cache_table[l1i_cpu_id][set][way].accessed = true;
+  }
+}
+
+bool l1i_is_accessed_timing_entry(uint64_t line_addr)
+{
+  uint32_t index = l1i_find_timing_mshr_entry(line_addr);
+  if (index < L1I_TIMING_MSHR_SIZE) {
+    return l1i_timing_mshr_table[l1i_cpu_id][index].accessed;
+  }
+  uint32_t set = line_addr % L1I_SET;
+  uint32_t way = l1i_find_timing_cache_entry(line_addr);
+  if (way < L1I_WAY) {
+    return l1i_timing_cache_table[l1i_cpu_id][set][way].accessed;
+  }
+  return false;
+}
+
+bool l1i_completed_request(uint64_t line_addr) { return l1i_find_timing_cache_entry(line_addr) < L1I_WAY; }
+
+bool l1i_ongoing_request(uint64_t line_addr) { return l1i_find_timing_mshr_entry(line_addr) < L1I_TIMING_MSHR_SIZE; }
+
+bool l1i_ongoing_accessed_request(uint64_t line_addr)
+{
+  uint32_t index = l1i_find_timing_mshr_entry(line_addr);
+  if (index == L1I_TIMING_MSHR_SIZE)
+    return false;
+  return l1i_timing_mshr_table[l1i_cpu_id][index].accessed;
+}
+
+uint64_t l1i_get_latency_timing_mshr(uint64_t line_addr)
+{
+  uint32_t index = l1i_find_timing_mshr_entry(line_addr);
+  if (index == L1I_TIMING_MSHR_SIZE)
+    return 0;
+  if (!l1i_timing_mshr_table[l1i_cpu_id][index].accessed)
+    return 0;
+  return l1i_get_latency(current_core_cycle[l1i_cpu_id], l1i_timing_mshr_table[l1i_cpu_id][index].timestamp);
+}
+
+// RECORD ENTANGLED TABLE
+
+uint32_t L1I_ENTANGLED_FORMATS[L1I_ENTANGLED_MAX_FORMATS] = {58, 28, 18, 13, 10, 8, 6};
+#define L1I_ENTANGLED_NUM_FORMATS 6
+
+uint32_t l1i_get_format_entangled(uint64_t line_addr, uint64_t entangled_addr)
+{
+  for (uint32_t i = L1I_ENTANGLED_NUM_FORMATS; i != 0; i--) {
+    if ((line_addr >> L1I_ENTANGLED_FORMATS[i - 1]) == (entangled_addr >> L1I_ENTANGLED_FORMATS[i - 1])) {
+      return i;
+    }
+  }
+  assert(false);
```

Comment:
This feels like a very dangerous assertion. There aren't guarantees that `(line_addr >> 58) == (entangled_addr >> 58)`.